### PR TITLE
Longer timeout for mapping update during resize

### PR DESCRIPTION
**Pull request by DaveCTurner**
We perform a mapping update when creating a resized index. Today we do this update using the regular put-mapping API with the default timeout of 30s, blocking the recovery thread while waiting for the response. In an overloaded cluster it can definitely take more than 30s to process this update, and that causes the recovery to fail. The only reasonable reactions to such a failure are either to retry (which moves us to the back of the master queue again) or to give up and leave the shard unassigned. It'd be better to just wait more patiently.

This commit extends the timeout to `TimeValue.MAX_VALUE`, moves to using the internal auto-put-mapping API, and moves to an async call to free up the thread for other work.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
I don't see a great way to write a failing test for this bug. I can for instance block cluster state updates until the old put-mapping request would have timed out, but that means the test will have to run for ≥30s. IMO the fact that the fix is the right shape, and still passes the existing test suite, is good enough.

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 1]
```diff
@@ -853,15 +858,20 @@ public IndexShard createShard(
...
+            client.execute(
+                clusterService.state().nodes().getMinNodeVersion().onOrAfter(Version.V_8_8_0)
+                    ? AutoPutMappingAction.INSTANCE
+                    : PutMappingAction.INSTANCE,
+                new PutMappingRequest().setConcreteIndex(shardRouting.index())
+                    .setConcreteIndex(shardRouting.index()) // concrete index - no name clash, it uses uuid
+                    .source(mapping.source().string(), XContentType.JSON)
+                    .timeout(TimeValue.MAX_VALUE)
+                    .masterNodeTimeout(TimeValue.MAX_VALUE),
+                new ThreadedActionListener<>(threadPool.generic(), listener.map(ignored -> null))
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/indices/IndicesService.java:873]: This slightly changes behavior during shutdown I think in that, we'd never complete the listener. I doubt that we wait for recoveries to complete anywhere in our shutdown sequence though, but wanted to be sure you had this on your radar too.

**Review by henningandersen — APPROVED** [review]
LGTM.

At a higher level, I wonder if we could move to a mechanism not requiring to put mappings here and rather do it either at shard started time or have a harder enforcement of read-only-ness. But that is for another day :slightly_smiling_face: .

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/indices/IndicesService.java:873]: ++ yes we fork to `GENERIC` all over the place during recovery already, see `org.elasticsearch.indices.recovery.RemoteRecoveryTargetHandler#executeRetryableAction`.

**Pull request merged**